src/commands/utils.rs

```diff
@@ -62,3 +71,93 @@ where
+///
+/// `candidate_coins` are the coins to consider for selection.
+///
+/// `base_tx` is the transaction to select coins for. It should be without any inputs
+/// and without a change output, but with all non-change outputs added.
+///
+/// `change_txo` is the change output to add if needed (with any value).
+///
+/// `feerate_vb` is the minimum feerate (in sats/vb).
+///
+/// `min_fee` is the minimum fee (in sats) that the selection must have.
+///
+/// `max_sat_weight` is the maximum size difference (in vb) of
+/// an input in the transaction before and after satisfaction.
+pub fn select_coins_for_spend(
+    candidate_coins: &[CandidateCoin],
+    base_tx: bitcoin::Transaction,
+    change_txo: bitcoin::TxOut,
+    feerate_vb: u64,
+    min_fee: u64,
+    max_sat_weight: usize,
+) -> Result<(Vec<Coin>, bitcoin::Amount), InsufficientFunds> {
+    let max_input_weight = TXIN_BASE_WEIGHT + max_sat_weight as u32;
+    let candidates: Vec<Candidate> = candidate_coins
+        .iter()
+        .map(|cand| Candidate {
+            input_count: 1,
+            value: cand.coin.amount.to_sat(),
+            weight: max_input_weight,
+            is_segwit: true, // We only support receiving on Segwit scripts.
+        })
+        .collect();
+    // Transaction base weight is calculated from transaction with no inputs and no change output.
+    let base_weight = base_tx.weight().to_wu() as u32;
+    let long_term_feerate = FeeRate::from_sat_per_vb(LONG_TERM_FEERATE_VB as f32);
+    let target = Target {
+        value: base_tx.output.iter().map(|o| o.value).sum(),
+        feerate: FeeRate::from_sat_per_vb(feerate_vb as f32),
+        min_fee,
+    };
+    let drain_weights = DrainWeights {
+        output_weight: {
+            let mut tx_with_change = base_tx;
+            tx_with_change.output.push(change_txo);
+            tx_with_change.weight().to_wu() as u32 - base_weight
+        },
+        spend_weight: max_input_weight,
+    };
+    // Change policy ensures any change output is not too small and that transaction waste is reduced.
+    let change_policy =
+        change_policy::min_value_and_waste(drain_weights, DUST_OUTPUT_SATS, long_term_feerate);
+
+    let mut selector = CoinSelector::new(&candidates, base_weight);
+    // Select any mandatory candidates.
+    for (i, cand) in candidate_coins.iter().enumerate() {
+        if cand.must_select {
+            selector.select(i);
+        }
+    }
+    if let Err(e) = selector.run_bnb(
+        LowestFee {
+            target,
+            long_term_feerate,
+            change_policy: &change_policy,
+        },
+        100_000,
+    ) {
+        warn!(
+            "Coin selection error: '{}'. Selecting coins by descending value per weight unit...",
+            e.to_string()
+        );
+        // If selection not possible by BnB, order coins by descending value ready for selection below.
+        selector.sort_candidates_by_descending_value_pwu();
+    }
```

Comment:
I'm not completely sure, but I suppose it's unlikely that BnB would fail (or at least not find a solution in 100,000 iterations) and selecting coins by descending value would succeed, so it should be fine to skip that step. We could re-add this fallback later if need be.

Reply:
It's not what i meant. The fallback is good and necessary, what i meant is it may only be necessary if we exhausted all the rounds, and not if the bnb returned early.

Reply:
Ahh, sorry didn't read it properly :) The BnB error will only be raised if it reaches the max number of rounds, so I don't think it can happen that it returns with an error early.

Reply:
Sorry, that's not completely right either :grimacing: It might return immediately if the candidates have insufficient funds, for example.

Reply:
Perhaps it's unlikely that it would ever need to use 100,000 rounds given the number of candidates we'll be dealing with, so any error would probably always be returned early in that case. But it may be simpler to attempt the fallback always rather than decide based on the number of rounds.

Reply:
Yeah, let's keep the fallback. It was just a nit as i was curious exactly about this point: we could just return InsufficientFunds early. But let's keep that for a potential followup. :)